src/miner_hbbft_sidecar.erl

```diff
@@ -0,0 +1,276 @@
+    gen_server:call(?SERVER, {set_group, Group}, infinity).
+
+new_round(Buf, Txns) ->
+    gen_server:call(?SERVER, {new_round, Buf, Txns}, infinity).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+init([]) ->
+    init(#state{});
+init(State) ->
+    case blockchain_worker:blockchain() of
+        undefined ->
+            erlang:send_after(500, self(), chain_check),
+            {ok, State#state{}};
+        Chain ->
+            Ledger = blockchain_ledger_v1:new_context(blockchain:ledger(Chain)),
+            Chain1 = blockchain:ledger(Ledger, Chain),
+            {ok, State#state{chain = Chain1}}
+    end.
+
+handle_call({set_group, Group}, _From, #state{group = OldGroup} = State) ->
+    lager:debug("setting group to ~p", [Group]),
+    case {OldGroup, Group} of
+        {undefined, undefined} ->
+            ok;
+        {P1, P2} when is_pid(P1) andalso is_pid(P2)  ->
+            ok;
+        {undefined, P} when is_pid(P) ->
+            ok = libp2p_swarm:add_stream_handler(blockchain_swarm:swarm(), ?TX_PROTOCOL,
+                                                 {libp2p_framed_stream, server,
+                                                  [blockchain_txn_handler, self(),
+                                                   fun(T) -> miner_hbbft_sidecar:submit(T) end]});
+        {P, undefined} when is_pid(P) ->
+            libp2p_swarm:remove_stream_handler(blockchain_swarm:swarm(), ?TX_PROTOCOL)
+    end,
+    {reply, ok, State#state{group = Group}};
+handle_call({submit, _}, _From, #state{group = undefined} = State) ->
+    lager:debug("submission with no group set"),
+    {reply, {error, no_group}, State};
+handle_call({submit, Txn}, From,
+            #state{chain = Chain,
+                   group = Group,
+                   queue = Queue,
+                   validations = Validations} = State) ->
+    Type = blockchain_txn:type(Txn),
+    lager:debug("got submission of txn: ~s", [blockchain_txn:print(Txn)]),
+    {ok, Height} = blockchain_ledger_v1:current_height(blockchain:ledger(Chain)),
+    case lists:member(Type, ?SlowTxns) of
+        true ->
+            Limit = application:get_env(miner, sidecar_parallelism_limit, 3),
+            case maps:size(Validations) of
+                N when N >= Limit ->
+                    Queue1 = Queue ++ [{From, Txn}],
+                    {noreply, State#state{queue = Queue1}};
+                _ ->
+                    {Attempt, V} = start_validation(Txn, From, Chain),
+                    {noreply, State#state{validations = Validations#{Attempt => V}}}
+            end;
+        false ->
+            case blockchain_txn:is_valid(Txn, Chain) of
+                ok ->
+                    case blockchain_txn:absorb(Txn, Chain) of
+                        ok ->
+                            catch libp2p_group_relcast:handle_command(Group, Txn),
```

Comment:
this is synchronous so if the hbbft handler is calling us with a new buffer at the same time, we can deadlock here?